processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java

```diff
@@ -0,0 +1,154 @@
+  {
+  };
+
+  @Override
+  public <T extends LogicalSegment> List<T> filterSegments(DataSourceMetadataQuery query, List<T> segments)
+  {
+    if (segments.size() <= 1) {
+      return segments;
+    }
+
+    final T min = segments.get(0);
+    final T max = segments.get(segments.size() - 1);
+
+    return Lists.newArrayList(
+        Iterables.filter(
+            segments,
+            new Predicate<T>()
+            {
+              @Override
+              public boolean apply(T input)
+              {
+                return (min != null && input.getInterval().overlaps(min.getInterval())) ||
+                       (max != null && input.getInterval().overlaps(max.getInterval()));
+              }
+            }
+        )
+    );
+  }
+
+  @Override
+  public QueryRunner<Result<DataSourceMetadataResultValue>> mergeResults(
+      final QueryRunner<Result<DataSourceMetadataResultValue>> runner
+  )
+  {
+    return new BySegmentSkippingQueryRunner<Result<DataSourceMetadataResultValue>>(runner)
+    {
+      @Override
+      protected Sequence<Result<DataSourceMetadataResultValue>> doRun(
+          QueryRunner<Result<DataSourceMetadataResultValue>> baseRunner,
+          Query<Result<DataSourceMetadataResultValue>> input,
+          Map<String, Object> context
+      )
+      {
+        DataSourceMetadataQuery query = (DataSourceMetadataQuery) input;
+        return Sequences.simple(
+            query.mergeResults(
+                Sequences.toList(
+                    baseRunner.run(query, context),
+                    Lists.<Result<DataSourceMetadataResultValue>>newArrayList()
+                )
+            )
+        );
+      }
+    };
+  }
+
+  @Override
+  public Sequence<Result<DataSourceMetadataResultValue>> mergeSequences(Sequence<Sequence<Result<DataSourceMetadataResultValue>>> seqOfSequences)
+  {
+    return new OrderedMergeSequence<>(getOrdering(), seqOfSequences);
+  }
+
+  @Override
+  public Sequence<Result<DataSourceMetadataResultValue>> mergeSequencesUnordered(Sequence<Sequence<Result<DataSourceMetadataResultValue>>> seqOfSequences)
+  {
+    return new MergeSequence<>(getOrdering(), seqOfSequences);
+  }
+
+  @Override
+  public ServiceMetricEvent.Builder makeMetricBuilder(DataSourceMetadataQuery query)
+  {
+    return new ServiceMetricEvent.Builder()
+        .setUser2(DataSourceUtil.getMetricName(query.getDataSource()))
+        .setUser4(query.getType())
+        .setUser6("false");
+  }
+
+  @Override
+  public Function<Result<DataSourceMetadataResultValue>, Result<DataSourceMetadataResultValue>> makePreComputeManipulatorFn(
+      DataSourceMetadataQuery query, MetricManipulationFn fn
+  )
+  {
+    return Functions.identity();
+  }
+
+  @Override
+  public TypeReference<Result<DataSourceMetadataResultValue>> getResultTypeReference()
+  {
+    return TYPE_REFERENCE;
+  }
+
+  @Override
+  public CacheStrategy getCacheStrategy(DataSourceMetadataQuery query)
+  {
+    return null;
```

Comment:
the maxIngestedEventTime changes with every new event ingested, probably it would make sense to make it cacheable when we start exposing more info from this query,